src/ol/controls/RoutingControl.js

```diff
@@ -0,0 +1,359 @@
+        };
+
+        viaPointFeatures.forEach((feature) => {
+          feature.on('change', this.relocateViaPoint);
+        });
+
+        // Write object with modify info
+        this.initialRouteDrag = {
+          viaPointFeatures,
+          oldRoute: route && route.clone(),
+          coordinate: e.mapBrowserEvent.coordinate,
+        };
+      });
+
+      this.modify.on('modifyend', (e) => {
+        /* Get the index for new viaPoint */
+        let newViaIndex;
+        const {
+          oldRoute,
+          relocateViaPoint,
+          viaPointFeatures,
+        } = this.initialRouteDrag;
+
+        // Unlisten feature change listeners
+        viaPointFeatures.forEach((feature) => {
+          feature.un('change', this.relocateViaPoint);
+        });
+
+        // If viaPoint is being relocated overwrite the old viaPoint
+        if (relocateViaPoint) {
+          return this.addViaPoint(
+            e.mapBrowserEvent.coordinate,
+            relocateViaPoint.get('index'),
+            1,
+          );
+        }
+
+        // In case there is no route overwrite first coordinate
+        if (!oldRoute) {
+          return this.addViaPoint(e.mapBrowserEvent.coordinate, 0, 1);
+        }
+
+        // Get the closest point from viaPoints to original route
+        const viaPointClosestPoints = this.viaPoints.map((viaPoint) => {
+          const snapPoint = oldRoute.getGeometry().getClosestPoint(viaPoint);
+          return snapPoint;
+        });
+
+        // Calculate the distance from the click point to each segment
+        let distanceToSegment;
+        for (let i = 0; i < viaPointClosestPoints.length - 1; i += 1) {
+          // Create a segment for each pair of (closest) viaPoints using turf lineSlice
+          const turfLineSegment = lineSlice(
+            createTurfPoint(viaPointClosestPoints[i]),
+            createTurfPoint(viaPointClosestPoints[i + 1]),
+            createTurfLine(oldRoute.getGeometry().getCoordinates()),
+          );
+
+          /* Create an ol LineString from segment and get the distance between
+           * the click point and the closest point on segment
+           */
+          const segment = new LineString(turfLineSegment.geometry.coordinates);
+          const distanceToClick = getDistance(
+            segment.getClosestPoint(this.initialRouteDrag.coordinate),
+            this.initialRouteDrag.coordinate,
+          );
+
+          // Return segment index with smallest distance to click point
+          if (!distanceToSegment || distanceToClick < distanceToSegment) {
+            distanceToSegment = distanceToClick;
+            newViaIndex = i + 1;
+          }
+        }
+
+        // Insert new viaPoint at given index
+        return this.addViaPoint(e.mapBrowserEvent.coordinate, newViaIndex);
+      });
+
+      // Add control ressources to map
+      this.map.addLayer(this.routingLayer);
+      this.onMapClickKey = this.map.on('singleclick', (e) => {
```

Comment:
use callback functions it will make the code more readable. this.onMapClick, this.onModifyStart....

Reply:
done